plugins/catalog-backend/src/processing/DefaultCatalogProcessingEngine.ts

```diff
@@ -33,34 +33,85 @@ import {
 } from './types';
 import { Stitcher } from '../stitching/Stitcher';
 import { startTaskPipeline } from './TaskPipeline';
+import { PluginTaskScheduler } from '@backstage/backend-tasks';
+import { Config } from '@backstage/config';
 
 const CACHE_TTL = 5;
 
 export type ProgressTracker = ReturnType<typeof progressTracker>;
 
 export class DefaultCatalogProcessingEngine implements CatalogProcessingEngine {
+  private readonly config: Config;
+  private readonly scheduler?: PluginTaskScheduler;
+  private readonly logger: Logger;
+  private readonly processingDatabase: ProcessingDatabase;
+  private readonly orchestrator: CatalogProcessingOrchestrator;
+  private readonly stitcher: Stitcher;
+  private readonly createHash: () => Hash;
+  private readonly pollingIntervalMs: number;
+  private readonly orphanCleanupIntervalMs: number;
+  private readonly onProcessingError?: (event: {
+    unprocessedEntity: Entity;
+    errors: Error[];
+  }) => Promise<void> | void;
+  private readonly tracker: ProgressTracker;
+
   private stopFunc?: () => void;
 
-  constructor(
-    private readonly logger: Logger,
-    private readonly processingDatabase: ProcessingDatabase,
-    private readonly orchestrator: CatalogProcessingOrchestrator,
-    private readonly stitcher: Stitcher,
-    private readonly createHash: () => Hash,
-    private readonly pollingIntervalMs: number = 1000,
-    private readonly onProcessingError?: (event: {
+  constructor(options: {
+    config: Config;
+    scheduler?: PluginTaskScheduler;
+    logger: Logger;
+    processingDatabase: ProcessingDatabase;
+    orchestrator: CatalogProcessingOrchestrator;
+    stitcher: Stitcher;
+    createHash: () => Hash;
+    pollingIntervalMs?: number;
+    orphanCleanupIntervalMs?: number;
+    onProcessingError?: (event: {
       unprocessedEntity: Entity;
       errors: Error[];
-    }) => Promise<void> | void,
-    private readonly tracker: ProgressTracker = progressTracker(),
-  ) {}
+    }) => Promise<void> | void;
+    tracker?: ProgressTracker;
+  }) {
+    this.config = options.config;
+    this.scheduler = options.scheduler;
+    this.logger = options.logger;
+    this.processingDatabase = options.processingDatabase;
+    this.orchestrator = options.orchestrator;
+    this.stitcher = options.stitcher;
+    this.createHash = options.createHash;
+    this.pollingIntervalMs = options.pollingIntervalMs ?? 1_000;
+    this.orphanCleanupIntervalMs = options.orphanCleanupIntervalMs ?? 30_000;
+    this.onProcessingError = options.onProcessingError;
+    this.tracker = options.tracker ?? progressTracker();
+
+    this.stopFunc = undefined;
+  }
 
   async start() {
     if (this.stopFunc) {
       throw new Error('Processing engine is already started');
     }
 
-    this.stopFunc = startTaskPipeline<RefreshStateItem>({
+    const stopPipeline = this.startPipeline();
```

Comment:
This confuses me, I guess startPipeline returns a function that would stop the pipeline?

Reply:
Exactly.